Add Zug consensus protocol

afck opened the pull request:
This allows choosing one of two consensus protocols in the chainspec, Zug or Highway.

Zug is a very simple leader-based partially synchronous protocol:
* In every round, the leader proposes a block (with some hash `h`). If a quorum of validators signs `Echo(h)`, it becomes that round's unique accepted candidate. If a quorum signs `Vote(true)`, the block and all its ancestors get finalized.
* If a quorum signs `Vote(false)`, the next round's leader is allowed to propose a block whose parent is _not_ `h`. If there is a quorum of `Echo(h)`, the next round's leader is allowed to propose a block whose parent _is_ `h`. (Both can be the case!)
* Validators vote `true` only after seeing a quorum of `Echo`es with matching hash. They vote `false` if they time out waiting for that.

A detailed description is in [this paper](https://arxiv.org/abs/2205.06314), using the gossiping variant (i.e. section 6.2, not 6.1).

Instead of the Highway round exponent, a more explicit `minimum_block_time` can now be configured and applies to both protocols.

There are still a few TODOs, and the tests should be expanded, but it has been successfully tried out in a local network.

Closes  https://github.com/casper-network/Governance/issues/310.

afck committed:
Prototype for a simple broadcast-and-vote consensus protocol.

afck committed:
Simplify and add some comments.

afck committed:
Deduplicate leader sequence.

Extract the leader sequence calculation so both `HighwayProtocol` and
`SimpleConsensus` can use it.

afck committed:
Add Fault type to SimpleConsensus.

This distinguishes between direct evidence and evidence from other eras.

afck committed:
Detect double-signing.

`SimpleConsensus` now detects if a validator signed two conflicting
proposals, echos or votes, broadcasts the information and also checks
whether the fault tolerance threshold was exceeded.

afck committed:
Don't propose switch block descendants.

afck committed:
Use corrected timestamp in proposed blocks.

afck committed:
Periodically sync protocol state.

SamuelSchlesinger committed:
Fixed some compilation errors due to private interfaces leaking through DataSize derivation

afck committed:
Fix has_voted; stop after switch block.

afck committed:
Merge remote-tracking branch 'public/dev' into simple_consensus

SamuelSchlesinger committed:
Extracted a few common parts between highway and simple consensus out

afck committed:
Update equivocators TODO.

afck committed:
Add participation logging; fix warnings.

afck committed:
Standstill alert

SamuelSchlesinger committed:
Fixed TODOs

afck committed:
Merge pull request #49 from casper-network/samschles/protocols_common

Protocols common code

SamuelSchlesinger committed:
Documented some things

SamuelSchlesinger committed:
Update node/src/components/consensus/protocols/simple_consensus.rs

Co-authored-by: Andreas Fackler <afck@users.noreply.github.com>

SamuelSchlesinger committed:
Update node/src/components/consensus/protocols/simple_consensus.rs

Co-authored-by: Andreas Fackler <afck@users.noreply.github.com>

SamuelSchlesinger committed:
Update node/src/components/consensus/protocols/simple_consensus.rs

Co-authored-by: Andreas Fackler <afck@users.noreply.github.com>

SamuelSchlesinger committed:
Update node/src/components/consensus/protocols/simple_consensus.rs

Co-authored-by: Andreas Fackler <afck@users.noreply.github.com>

SamuelSchlesinger committed:
Fixed comment on contains and removed redundant TODO on let decl for Params in simple_consensusC

SamuelSchlesinger committed:
Merge branch 'samschles/document_simple_consensus' of github.com:casper-network/casper-node-private into samschles/document_simple_consensus

SamuelSchlesinger committed:
Fixed more bits of simple_consensus documentation

SamuelSchlesinger committed:
Update node/src/components/consensus/protocols/simple_consensus.rs

Co-authored-by: Andreas Fackler <afck@users.noreply.github.com>

SamuelSchlesinger committed:
Merge branch 'samschles/document_simple_consensus' of github.com:casper-network/casper-node-private into samschles/document_simple_consensus

SamuelSchlesinger committed:
Update node/src/components/consensus/protocols/simple_consensus.rs

Co-authored-by: Andreas Fackler <afck@users.noreply.github.com>

SamuelSchlesinger committed:
Merge pull request #50 from casper-network/samschles/document_simple_consensus

Documented some things in simple consensus

SamuelSchlesinger committed:
Fixed broken build -- whoops

afck committed:
Recompute quorums when faults are detected.

afck committed:
Make sync message smaller, use bit fields.

Sync messages now use a bit field for a subset of validators in only a
single round to tell a peer about that part of the sender's protocol
state. However, independent of whether a quorum has been reached, it
always lists the individual validators in that set from which a vote
or echo has been received. Unlike the previous approach this guarantees
that everyone will eventually either see a quorum of "false" votes, or a
quorum of echos, even if the correct validators are not fully connected.

afck committed:
Fix validator_bit_field functions.

afck committed:
Merge remote-tracking branch 'public/dev' into simple_consensus

afck committed:
Add comments about bit fields.

afck committed:
Address review comments.

afck committed:
Address review comment.

afck committed:
Add comments to Message::SyncState.

afck committed:
Merge pull request #54 from casper-network/afck_simple_consensus

Recompute quorums; make sync messages smaller.

afck committed:
Move most protocol logic into update; memoize current round.

afck committed:
Add unit tests for SimpleConsensus.

afck committed:
Merge remote-tracking branch 'public/dev' into upgrade2

afck committed:
Merge pull request #73 from casper-network/merge_dev_into_simple_consensus

Merge dev into simple consensus

afck committed:
Merge pull request #72 from casper-network/update

Move most protocol logic into update; memoize current round.

afck committed:
Respect min round length and era start time when proposing.

This moves the proposal logic into `update` as well, and delays
proposals such that the minimum round length between blocks and the era
start time are respected.

afck committed:
Fix is_quorum.

afck committed:
Don't call update recursively.

afck committed:
Merge pull request #75 from casper-network/afck_sc_unit_tests

Add unit tests for SimpleConsensus.

afck committed:
Detect and report inactive validators.

afck committed:
Merge branch 'simple_consensus' into afck_proposal_delay

afck committed:
Fix Clippy issues; proposer is never inactive.

afck committed:
Add some comments and extract create_round.

afck committed:
Fix SimpleConsensus test: take new timestamp rule into account.

afck committed:
Check inactive claims before accepting proposals.

afck committed:
Update tests.

afck committed:
Merge pull request #76 from casper-network/afck_proposal_delay

Respect min round length and era start time when proposing.

afck committed:
Extract register_activity.

afck committed:
Call update also when a validator becomes active.

When a validator becomes active that may make a pending proposal
accepted, so `update` needs to be called.

This commit replaces `update`'s argument with a field that marks a round
as dirty, and sets that field also when validators become active.

afck committed:
Add some module documentation.

afck committed:
Merge pull request #77 from casper-network/afck_inactive

Detect and report inactive validators.

afck committed:
Add tests for validator bit fields.

afck committed:
Add a test for is_quorum and quorum_threshold.

Commit:
Pause mode

temp

Commit:
Increase timeout each time we update without commiting a new round

Commit:
Addressed review comments

Commit:
Removed base_proposal_timeout because it is not necessary to memoize the round length inferred from a previous era after a round is committed

Commit:
Addressed PR comments

afck committed:
Merge pull request #78 from casper-network/afck_test_validator_bit_field

Add tests for bit fields, and quorum.

SamuelSchlesinger committed:
Merge pull request #80 from casper-network/samschles/timeout-increase-transform

Increase Round Timeout until Commit

SamuelSchlesinger committed:
Merge pull request #79 from casper-network/samschles/pause

Pause mode

afck committed:
Fix handling of timestamps and proposals.

Pass in timestamps instead of calling `Timestamp::now()`.

Make sure a call to `update` is scheduled if a proposal cannot be made
yet due to minimum round length.

Extend a unit test to cover proposals.

afck committed:
Test behavior after switch block.

Proposals after the switch block must be empty.

afck committed:
Test detecting if FTT is exceeded.

afck committed:
Pass in RNG instead of using thread_rng.

afck committed:
Test standstill detection.

afck committed:
Merge pull request #82 from casper-network/afck_rng

Pass in RNG instead of using thread_rng.

afck committed:
Merge remote-tracking branch 'upstream/simple_consensus' into afck_proposal_test

afck committed:
Add RNG to new handle_timer calls.

afck committed:
Merge pull request #81 from casper-network/afck_proposal_test

Add SimpleConsensus unit tests, minor fixes.

afck committed:
Merge remote-tracking branch 'public/dev' into afck_merge_dev

afck committed:
Merge pull request #83 from casper-network/afck_merge_dev

Merge dev into simple_consensus.

afck committed:
Don't add faulty echos/votes after handle_faulty; filter faulty in reply to SyncState.

If we detect double-signing we remove all the echos and votes of the
faulty validator, but we used to still add the second of the two
conflicting signatures afterwards. That was a bug.

We now also don't send echos and votes of faulty validators when
responding to SyncState messages, since these are redundant.

afck committed:
Add a unit test for sending SyncState.

afck committed:
Treat SyncState with wrong hash like one without; add test.

If someone sends us a `SyncState` message with a proposal hash that
doesn't match the one for which we already have a quorum of echos, we
should send them the echos for the winning proposal.

afck committed:
Add Params struct for simple consensus.

This moves all constant values - i.e. values that don't change for the
duration of one era - into a `Params` struct.

afck committed:
Add a SimpleConsensus config section.

afck committed:
Add consensus protocol choice to chainspec.

afck committed:
Add minimum block time to chainspec.

afck committed:
Add changelog entry.

afck committed:
Move finality_threshold_fraction from Highway to core config.

afck committed:
Move max_execution_delay from Highway to consensus config.

afck committed:
Merge pull request #84 from afck/test_sync_state

Test (and fix) SyncState handling in simple consensus.

afck committed:
Fix missed itst13.config.toml file.

afck committed:
Make TimeDiff config options non-optional.

afck committed:
Merge pull request #85 from afck/simple_consensus_params

Simple consensus params and config options.

afck committed:
Clean up consensus error handling.

This removes the `InvalidIncomingMessage` variant from
`ProtocolOutcomes` because it does the same as `Disconnect`, plus some
logging. We log the details inside the consensus implementations
instead, where more information is available.

Now we can also make `ProtocolOutcome` implement `PartialEq`, which
simplifies some unit tests.

afck committed:
Merge pull request #86 from casper-network/simple_consensus_error

Clean up consensus error handling.

afck committed:
Merge branch 'dev' into 'simple_consensus'

afck committed:
Extract simple_consensus submodules.

afck committed:
Hide Round fields, add utility methods.

Make the fields of `Round` and `RoundOutcome` private and add some
utility methods.

afck committed:
Remove proposal signatures; use echo signatures instead.

afck committed:
Fix missed test chainspec files.

afck committed:
Merge pull request #87 from afck/merge_dev

Merge dev into simple_consensus

afck committed:
Add a Message::Evidence variant.

This allows us to respect the `evidence_only` mode.

afck committed:
Vote immediately to skip a faulty leader's round.

If we already know that the current round leader is faulty, don't wait
for the timeout and immediately sign a `Vote(false)`.

afck committed:
Deduplicate timers; refactor update.

This adds a field to `SimpleConsensus` so it remembers the earliest
timer it scheduled, and avoids scheduling them again.

It also refactors `update` to make it a bit more readable, and removes a
redundant timer from `update_accepted_proposal`.

afck committed:
Merge pull request #88 from afck/evidence

Make evidence more explicit, respect evidence_only mode

afck committed:
Deduplicate proposal-with-echo creation.

afck committed:
Store leader index in Round.

afck committed:
Sync information about inactive validators.

`SyncState` now includes a list of validators seen as active. The peer
responds with evidence for activity of missing entries, i.e. with one of
their `SignedMessage`s.

afck committed:
Prune orphaned and finalized rounds.

afck committed:
Fix DataSize for SimpleConsensus.

afck committed:
Also prune wrong echoes from finalized rounds.

afck committed:
Update comment.

afck committed:
Address review comments.

afck committed:
Merge pull request #89 from afck/prune

Prune orphaned and finalized rounds.

afck committed:
Merge 'feat-2.0' into 'simple_consensus'

afck committed:
Fix casper-types  build with no-std.

afck committed:
Merge pull request #90 from afck/merge-feat-2.0

Merge feat-2.0 into simple_consensus

afck committed:
Never send a message we didn't add to the protocol state.

That would put us in danger of double-signing.

afck committed:
Don't prune signatures.

Removing signatures is a bit dangerous: it makes it easy to add
double-signing bugs in the future. And they are small anyway, so not
much memory is wasted.

Instead, just filter out unnecessary ones in `handle_sync_state`.

Commit:
Added a write ahead log for zug to allow for fault tolerance.

afck committed:
Fix upgrade test config and chainspec.

Joining nodes can only catch up if the block time is much longer than
the sync state interval.

afck committed:
Remove outdated comment.

What was decided regarding proposals with future timestamps is already
implemented, so the TODO is outdated.

casperlabs-bors-ng[bot] committed:
Merge #91

91: Fix create_message. r=afck a=afck

Fixes a double-signing bug that was caused by pruning. `create_message` would send messages that it didn't add to the protocol state.

As a precaution, we also stop pruning signatures, and instead just filter out unnecessary ones in `handle_sync_state`.

To fix the upgrade tests we configure the sync state interval to be much shorter than the block time, so that joining nodes have a chance to catch up.

Closes https://github.com/casper-network/Governance/issues/329, https://github.com/casper-network/Governance/issues/242.


Co-authored-by: Andreas Fackler <andreas@casperlabs.io>

afck committed:
Merge remote-tracking branch 'feat-2.0' into 'zug'

casperlabs-bors-ng[bot] committed:
Merge #92

92: Merge feat-2.0 into zug r=afck a=afck

Merge `feat-2.0` into `zug`.

https://github.com/casper-network/Governance/issues/310


Co-authored-by: Rafał Chabowski <rafal@casperlabs.io>
Co-authored-by: casperlabs-bors-ng[bot] <82463608+casperlabs-bors-ng[bot]@users.noreply.github.com>
Co-authored-by: Mateusz <mateusz@casperlabs.io>
Co-authored-by: Andreas Fackler <andreas@casperlabs.io>

afck committed:
Suppress fewer Clippy lints, make arithmetic safe.

Remove some `allow` attributes, especially regarding integer arithmetic.
Instead, handle all cases where it could over- or underflow, so that in
theory up to 2^32 - 1 validators are supported.

afck committed:
Rename 'Simple' to 'Zug' in the consensus chainspec option.

afck committed:
Use configured proposal timeout; always set proposal timer

afck committed:
Annotate bit field type.

casperlabs-bors-ng[bot] committed:
Merge #93

93: Make arithmetic safe, rename chainspec option to Zug. r=afck a=afck

This removes some clippy lint suppressions and correctly handles arithmetic overflows in quorum and validator index calculations. It also renames the `Simple` consensus protocol chainspec option to `Zug`, but doesn't rename all identifiers in the code yet. (The rest of https://github.com/casper-network/Governance/issues/333 will be done when there are fewer merge conflicts.)

Closes https://github.com/casper-network/Governance/issues/236.

Co-authored-by: Andreas Fackler <andreas@casperlabs.io>

afck committed:
Restart current round after unpausing.

This resets `self.current_timeout` after unpausing. The effect is both
that the current round's proposer immediately creates another proposal
(unless they've already proposed), and that the other validators restart
their timeout, so there is a chance the proposal will get accepted.

In nctl tests with block time 0 the pause mode is often activated and
deactivated in quick succession, which caused a lot of proposals getting
dropped. This change fixes that problem.

afck committed:
Fix a new Clippy warning in consensus.

https://rust-lang.github.io/rust-clippy/master/index.html#derive_partial_eq_without_eq

casperlabs-bors-ng[bot] committed:
Merge #95

95: Restart current round after unpausing. r=afck a=afck

This resets `self.current_timeout` after unpausing. The effect is both that the current round's proposer immediately creates another proposal (unless they've already proposed), and that the other validators restart their timeout, so there is a chance the proposal will get accepted.

In nctl tests with block time 0 the pause mode is often activated and deactivated in quick succession, which caused a lot of proposals getting dropped. This change fixes that problem.

Closes https://github.com/casper-network/Governance/issues/342.

Co-authored-by: Andreas Fackler <andreas@casperlabs.io>

afck committed:
Merge remote-tracking branch 'feat-2.0' into 'merge-feat-2.0'

casperlabs-bors-ng[bot] committed:
Merge #96

96: Merge feat-2.0 into zug r=afck a=afck

Merge `feat-2.0` into `zug`.

Conflicts were in `node/CHANGELOG.md` and `node/src/components/consensus/utils.rs`.

https://github.com/casper-network/Governance/issues/310


Co-authored-by: Mateusz <mateusz@casperlabs.io>
Co-authored-by: Fraser Hutchison <fraser@casperlabs.io>
Co-authored-by: George Pisaltu <georgep@casperlabs.io>
Co-authored-by: Bartłomiej Kamiński <bart@casperlabs.io>
Co-authored-by: Andreas Fackler <andreas@casperlabs.io>
Co-authored-by: casperlabs-bors-ng[bot] <82463608+casperlabs-bors-ng[bot]@users.noreply.github.com>

Commit:
Fixed some review comments

Commit:
Added corruption recovery and other pleasantries like working tests

Commit:
Fixed some more things, added some tests of the read functionality -- currently not working so well

Commit:
Fixed error in decoding, running into problem with the consumption of the ReadWAL.

afck committed:
Better proposal timeout adaptation.

Instead of just doubling it every round and resetting it whenever a
block gets finalized, adapt the proposal timeout more slowly, and in a
way that makes it settle roughly at a configurable multiple of a certain
percentile of measured delays.

afck committed:
Remove ValidatorID bound again; fix a few warnings

afck committed:
Call record_entry everywhere where insert_proposal is called.

afck committed:
Remove ConsensusConstructor type.

afck committed:
Simplify unit test.

afck committed:
Simplify WAL.

afck committed:
Don't activate as validator without WAL file.

afck committed:
Simplify test.

afck committed:
If passed in, open WAL in activate_validator.

afck committed:
Return false from record_entry if no WAL.

afck committed:
Add comments; never activate when faulty.

afck committed:
Admit if we're faulty.

afck committed:
Address review comments.

Commit:
Added a test for the unhappy path

afck committed:
Address review comments.

Commit:
Added newline in wal.rs test

afck committed:
Fix proposal timeout; add a unit test.

casperlabs-bors-ng[bot] committed:
Merge #98

98: Better proposal timeout adaptation. r=afck a=afck

Instead of just doubling it every round and resetting it whenever a block gets finalized, adapt the proposal timeout more slowly, and in a way that makes it settle roughly at a configurable multiple of a certain percentile of measured delays.

Closes https://github.com/casper-network/Governance/issues/334.


Here's the original proposal, but note that I've made slightly different choices for how to represent the parameters:


## Motivation

In Zug, at the start of every round, all validators set a timer: If no proposal gets accepted before it fires, they vote to allow skipping the round. The timeout should not be too long: A single faulty leader will stall the network until the timer fires. But if it is too short, some correct leaders' blocks will get orphaned because they don't manage to get them accepted in time.

We do want the timeout to dynamically adapt, but it needs to be Byzantine fault tolerant: With a configured fault tolerance threshold _f_, we have to limit the extent to which that percentage of nodes can influence it. That is why a median or percentile is more useful than a mean, because the latter reacts strongly to single outliers. The timeout should roughly be based on some _X_-th percentile of measured proposal delays, between _f_ and _100% - f_ of the validators. (_X = 50%_ would be the median.)

Since the actual delay to proposal acceptance already depends on the fastest quorum anyway, i.e. the fastest _(100% + f) / 2_ of validators, it makes sense to use that as the value _X_.

Instead of keeping track of a rolling window of measurements and computing a percentile, we propose something simpler with similar properties:


## Proposal

There is a configurable initial proposal timeout value, but every time a new round begins it gets updated. A grace period factor _G_ (default 3) and an inertia _R_ (default 10) are also configurable.

* We _increase_ the timeout by multiplying it with a factor _2<sup>1 / (R (1 - f))</sup>_ whenever
    * we vote `false` (i.e. our timer fired without an accepted proposal), _or_
    * we vote `true` (i.e. we accept a proposal in time) after a time _t_ where _G t_ is greater than our current proposal timeout.
* Whenever we vote `true` (i.e. we accept a proposal in time) after a time _t_ where _G t_ is less or equal to our current proposal timeout, we _decrease_ it, dividing it by _2<sup>1 / (R (1 + f))</sup>_.
* _But_: When increasing whe never go higher than _G t_, and when decreasing whe never go lower than _G t_.

In other words whenever we increase the timeout, we add _1 / (R (1 - f))_ to its binary logarithm, and when we decrease it we subtract _1 / (R (1 + f))_. So:

* If our timeout is way too low and all proposals are too slow, the timeout will double after _~ R (1 - f)_ blocks.
* If our timeout is way too high and all proposals get accepted quickly, it will halve after _~ R (1 + f)_ blocks.
* If our timeout is roughly _G_ times as long as it takes the slowest validator among the fastest quorum to get a proposal accepted, then a fraction of _(1 - f) / 2_ rounds will cause an increase and _(1 + f) / 2_ a decrease, so the increases and decreases will cancel out over time.

So the timeout will settle at a value where, if you want to get your proposal accepted, you shouldn't be slower by a factor of _≥ G_ compared to the slowest validator among the fastest _(100% + f) / 2_.

(Note that if we move to round robin, where everyone gets to be leader equally often, we need to multiply both exponents with _w / a_, where _w_ is the proposer's weight, and _a_ is the average weight.)

Co-authored-by: Andreas Fackler <andreas@casperlabs.io>

afck committed:
Merge remote-tracking branch 'public/feat-2.0' into merge

afck committed:
Add new settings to the new nctl config files.

casperlabs-bors-ng[bot] committed:
Merge #101

101: Merge feat-2.0 into zug r=afck a=afck

The only merge conflict was in the changelog, but I had to fix two missed nctl test chainspec and nine new nctl test config files.

https://github.com/casper-network/Governance/issues/310

Co-authored-by: Fraser Hutchison <fraser@casperlabs.io>
Co-authored-by: casperlabs-bors-ng[bot] <82463608+casperlabs-bors-ng[bot]@users.noreply.github.com>
Co-authored-by: TomVasile <43349666+TomVasile@users.noreply.github.com>
Co-authored-by: Bartłomiej Kamiński <bart@casperlabs.io>
Co-authored-by: Marc Brinkmann <marc@casperlabs.io>
Co-authored-by: George Pisaltu <georgep@casperlabs.io>
Co-authored-by: Rafał Chabowski <rafal@casperlabs.io>
Co-authored-by: Mateusz <mateusz@casperlabs.io>

Commit:
Refactored test to improve clarity

casperlabs-bors-ng[bot] committed:
Merge #99

99: WAL for Zug r=afck a=SamuelSchlesinger

Created a WAL for Zug so that nodes have access to messages they've received and sent in the past locally. This helps nodes prevent double-signing, makes it easier to join the era as a participant after shutting down, and allows for the network to be fully shutdown without breaking the era.

Co-authored-by: Samuel Schlesinger <sam@casperlabs.io>
Co-authored-by: Andreas Fackler <andreas@casperlabs.io>

afck committed:
Merge 'feat-2.0' into 'zug'

afck committed:
Fix Clippy warning.

afck committed:
Rename 'simple consensus' to 'Zug'.

casperlabs-bors-ng[bot] committed:
Merge #102

102: Merge feat-2.0 into zug r=afck a=afck

https://github.com/casper-network/Governance/issues/310

Co-authored-by: Mateusz <mateusz@casperlabs.io>

casperlabs-bors-ng[bot] committed:
Merge #103

103: Rename "Simple Consensus" to "Zug". r=afck a=afck

This renames all identifiers, files and chainspec/config options to use the protocol implementation's new name, "Zug".
(Also fixes a Clippy warning in `wal.rs`.)

Closes https://github.com/casper-network/Governance/issues/333.

Co-authored-by: Andreas Fackler <andreas@casperlabs.io>

afck committed:
Rename SyncState to SyncRequest

afck committed:
Add a SyncResponse message.

This replaces the multiple individual messages that were sent as a
response to the sync request, and deduplicates their common data.

afck committed:
Extend sync test.

afck committed:
Use the demand mechanism for Zug requests.

afck committed:
Address review comment.

casperlabs-bors-ng[bot] committed:
Merge #105

105: Deduplicate sync response data and make it a single message. r=afck a=afck

Instead of sending lots of messages with the same instance ID, round number, hash, etc. in response to a `SyncRequest` (formerly `SyncState`), send a single `SyncResponse` with that data deduplicated.

Closes https://github.com/casper-network/Governance/issues/239.


Co-authored-by: Andreas Fackler <andreas@casperlabs.io>

afck committed:
Add basic block rewards.

This may still change in the future, but it is certainly better than no
rewards at all: It grants the full block reward to the proposer for now.

afck committed:
Memoize proposal hash.

This adds a new `HashedProposal` struct containing a proposal and its
hash, and avoids redundantly computing the hash multiple times.

afck committed:
Remove double serialization for consensus messages.

afck committed:
Improve logging: less spam, more cases.

This changes some log messages to use `Display` instead of `Debug`, and
omits some large data, to make them more readable.

It also adds logging in Zug for when a round becomes skippable, when it
actually gets skipped, and when a round becomes committed but is still
waiting for the accepted proposal.

afck committed:
Address review comments.

afck committed:
Address review comments.

afck committed:
Avoid redundant update loops; log new rounds.

This adds an info log message when a new round begins, and makes the
`update` function spend only one iteration in each round, resetting the
dirty round to `None` at the end.

Before, it would sometimes mark the round as dirty that it's currently
handling, because its own echoes or votes can create a new quorum.
However, `update_round` processes everything in the right order, so that
any new quorum will be handled immediately.

afck committed:
Improve log messages: prefer Display over Debug

afck committed:
Log a warning if a timer is delayed by more than one second.

casperlabs-bors-ng[bot] committed:
Merge #106

106: Use the demand mechanism for Zug requests; memoize proposal hash. r=afck a=afck

* This makes the network aware that `SyncRequest` is a request, so that the number of in-flight requests get limited by backpressure.
* It replaces the prototype block rewards with something reasonable, even though this may still change in the future.
* It memoizes the proposal hash computation.
* And it gets rid of the double serialization: The `Vec<u8>` in `ConsensusProtocol` and `ProtocolOutcome` are replaced with `EraMessage` and `EraRequest` enums. Those get serialized/deserialized only as part of `crate::protocol::Message`.
* Improves logging, and covers the cases where a round becomes skippable, gets skipped, and where it becomes committed while still waiting for the accepted proposal.

Closes https://github.com/casper-network/Governance/issues/240, https://github.com/casper-network/Governance/issues/243, https://github.com/casper-network/Governance/issues/362.


Co-authored-by: Andreas Fackler <andreas@casperlabs.io>

afck committed:
Merge 'feat-2.0' into 'zug'

casperlabs-bors-ng[bot] committed:
Merge #107

107: Merge feat-2.0 into zug. r=afck a=afck

No significant merge conflicts.

https://github.com/casper-network/Governance/issues/310

Co-authored-by: Bartłomiej Kamiński <bart@casperlabs.io>
Co-authored-by: Rafał Chabowski <rafal@casperlabs.io>
Co-authored-by: TomVasile <43349666+TomVasile@users.noreply.github.com>
Co-authored-by: Marc Brinkmann <marc@casperlabs.io>
Co-authored-by: Michał Papierski <michal@casperlabs.io>
Co-authored-by: casperlabs-bors-ng[bot] <82463608+casperlabs-bors-ng[bot]@users.noreply.github.com>

afck committed:
Merge 'feat-2.0' into 'zug'

afck committed:
Don't schedule outdated timers.

This passes both the scheduled and the actual current timestamp into
`ConsensusProtocol::handle_timer`.

Zug now schedules the periodic timers for the configured period from
_now_ instead of from the planned timestamp, to avoid generating even
more events when the scheduler queue is already too full.

casperlabs-bors-ng[bot] committed:
Merge #109

109: Don't schedule outdated timers. r=afck a=afck

This passes both the scheduled and the actual current timestamp into `ConsensusProtocol::handle_timer`.

Zug now schedules the periodic timers for the configured period from _now_ instead of from the planned timestamp, to avoid generating even more events when the scheduler queue is already too full.

This is related to https://github.com/casper-network/Governance/issues/364, although it is unlikely to be the root cause.


Co-authored-by: Andreas Fackler <andreas@casperlabs.io>

casperlabs-bors-ng[bot] committed:
Merge #108

108: Merge feat-2.0 into zug. r=afck a=afck

https://github.com/casper-network/Governance/issues/310

Co-authored-by: Michał Papierski <michal@casperlabs.io>
Co-authored-by: Karan Dhareshwar <karan@casperlabs.io>
Co-authored-by: TomVasile <43349666+TomVasile@users.noreply.github.com>
Co-authored-by: casperlabs-bors-ng[bot] <82463608+casperlabs-bors-ng[bot]@users.noreply.github.com>
Co-authored-by: Fraser Hutchison <fraser@casperlabs.io>
Co-authored-by: George Pisaltu <georgep@casperlabs.io>

afck committed:
Use minimum_block_time in Highway.

Instead of configuring a separate minimum round exponent, just use the
minimum_block_time as the minimum round length. Instead of a
millisecond, the minimum block time now corresponds to an exponent 0.

afck committed:
Update CHANGELOG.

afck committed:
Merge 'feat-2.0' into 'zug'

afck committed:
Fix long lines.

casperlabs-bors-ng[bot] committed:
Merge #112

112: Merge feat-2.0 into zug. r=afck a=afck

Only a trivial conflict in the changelog.

https://github.com/casper-network/Governance/issues/310

Co-authored-by: Rafał Chabowski <rafal@casperlabs.io>
Co-authored-by: casperlabs-bors-ng[bot] <82463608+casperlabs-bors-ng[bot]@users.noreply.github.com>
Co-authored-by: Fraser Hutchison <fraser@casperlabs.io>

afck committed:
Merge remote-tracking branch 'upstream/zug' into hw_min_block_time

afck committed:
Fix synchronizer test comments.

afck committed:
Add a comment about the maximum round exponent.

casperlabs-bors-ng[bot] committed:
Merge #111

111: Use minimum_block_time in Highway. r=afck a=afck

Instead of configuring a separate minimum round exponent, just use the `minimum_block_time` as the minimum round length. Instead of a millisecond, an exponent 0 now corresponds to the minimum block time.

Closes https://github.com/casper-network/Governance/issues/367.

Co-authored-by: Andreas Fackler <andreas@casperlabs.io>

afck committed:
Merge remote-tracking branch 'feat-2.0' into zug

afck committed:
Apply make format.

afck commented:
That seems to be flaky… not sure if that's a new thing. Retrying…

afck commented:
bors try

afck commented:
I can't reproduce this locally; `cargo test run_participating_network` always passes for me locally. Must be a timing issue then. :confused:

afck committed:
Increase run_participating_network timeout.

afck commented:
bors try

afck commented:
bors try

fizyk20 commented on node/src/components/consensus.rs:
```diff
@@ -49,23 +50,94 @@ use crate::{
 #[derive(DataSize, Clone, Serialize, Deserialize)]
 pub(crate) enum ConsensusMessage {
     /// A protocol message, to be handled by the instance in the specified era.
-    Protocol { era_id: EraId, payload: Vec<u8> },
```
I'm wondering if changing the type of `payload` is what we want - IIRC, we made it opaque on purpose, so that plugging in another consensus algorithm wouldn't require modifications here? But then again, we don't use that many consensus algorithms, and having a proper type here looks cleaner, so I think that's okay :+1:

afck replied:
Yes, we replaced the dynamic with static dispatch here, and if we add a third protocol we'll have to add another enum variant to `EraMessage`. I think this is preferable compared to the double serialization, and also makes debugging easier.

fizyk20 commented on node/src/components/consensus.rs:
```diff
@@ -94,9 +166,12 @@ pub struct ResolveValidity {
     /// An incoming network message.
+    #[from]
     Incoming(ConsensusMessageIncoming),
+    /// An incoming request message.
```
Should we stick to calling it a "demand"? TBH I don't remember what demands were right now, but describing them as "request messages" doesn't really clarify that much, and could introduce some mess in the terminology :sweat_smile:

fizyk20 commented on node/src/components/consensus.rs:
```diff
@@ -164,11 +239,34 @@ impl Display for ConsensusMessage {
+            Event::DemandIncoming(demand) => {
+                write!(
+                    f,
+                    "request from {:?}: {}",
```
Same here - should we keep the terminology consistent?

fizyk20 commented on node/src/components/consensus/config.rs:
```diff
@@ -20,26 +22,34 @@ use crate::{
     /// Highway-specific node configuration.
     pub(crate) highway: HighwayConfig,
+    /// Zug-specific node configuration.
+    pub(crate) zug: ZugConfig,
```
Should we also add `#[serde(default)]` to these? As it is now, I believe we'll have to configure both algorithms in the config, even if we're going to only use one?

fizyk20 commented on node/src/components/consensus/era_supervisor.rs:
```diff
@@ -450,18 +427,34 @@ impl EraSupervisor {
+                start_time,
+                seed,
+                now,
+                self.unit_file(&instance_id),
```
Wait... How is it possible that `Zug` requires the unit file, and `Highway` doesn't? :thinking:

afck replied:
Highway gets it in `activate_validator`.
Zug always stores the state, even in non-validators. (Won't be relevant in 2.0, but still makes sense conceptually, I think.)

fizyk20 commented on node/src/components/consensus/era_supervisor.rs:
```diff
@@ -640,6 +644,41 @@ impl EraSupervisor {
+        effect_builder: EffectBuilder<REv>,
+        rng: &mut NodeRng,
+        sender: NodeId,
+        demand: ConsensusRequestMessage,
```
`ConsensusDemand` is already taken elsewhere, isn't it? In any case, maybe the type name should be something with "demand" in it, too?

fizyk20 commented on node/src/components/consensus/era_supervisor/era.rs:
```diff
@@ -195,11 +190,19 @@ impl DataSize for Era {
                 } else {
                     (*highway).estimate_heap_size()
                 }
+            } else if let Some(zug) = any_ref.downcast_ref::<Zug<ClContext>>() {
+                if *CASPER_ENABLE_DETAILED_CONSENSUS_METRICS {
+                    let detailed = (*zug).estimate_detailed_heap_size();
+                    match serde_json::to_string(&detailed) {
+                        Ok(encoded) => debug!(%encoded, "consensus memory metrics"),
+                        Err(err) => warn!(%err, "error encoding consensus memory metrics"),
+                    }
+                    detailed.total()
+                } else {
+                    (*zug).estimate_heap_size()
+                }
```
Would it be possible to downcast to something like `&dyn DataSize` regardless of which consensus that is?

afck replied:
Unfortunately `DataSize` isn't object safe because of its associated `const`s.

fizyk20 commented on node/src/components/consensus/highway_core/finality_detector/rewards.rs:
```diff
@@ -136,7 +136,8 @@ fn round_participation<'a, C: Context>(
     maybe_unit.map_or(RoundParticipation::No, |(vh, unit)| {
-        if unit.round_exp > r_id.trailing_zeros() {
+        #[allow(clippy::integer_arithmetic)]
+        if r_id.millis() % unit.round_len.millis() != 0 {
```
Maybe add a comment that `round_len` cannot be 0 in the unit, because it is still defined via the exponent in the `WireUnit`s that are sent over the network? So that a malicious validator cannot make it be 0.

fizyk20 commented on node/src/components/consensus/protocols/highway.rs:
```diff
@@ -763,13 +748,10 @@ where
                         trace!("received an invalid vertex");
```
If we add a `warn!` below, does this `trace!` still make sense?

fizyk20 reviewed (commented):
I've only skimmed through the code of Zug itself, as I expect it's the same code that already was reviewed during its development? If there are some changes that should be looked at, I can revisit it.

Anyway, looks good, just a few nitpicks :wink:

fizyk20 commented on node/src/components/consensus/protocols/zug.rs:
```diff
@@ -0,0 +1,2218 @@
+        prev_cp: Option<&dyn ConsensusProtocol<C>>,
+        era_start_time: Timestamp,
+        seed: u64,
+        _now: Timestamp,
```
Since we removed the `ConsensusConstructor` type and `Zug::new_boxed` is now called directly, maybe this could be removed? Or is it still used somewhere?

fizyk20 commented on node/src/components/consensus/protocols/zug.rs:
```diff
@@ -0,0 +1,2218 @@
+    /// Returns whether the switch block has already been finalized.
+    fn finalized_switch_block(&self) -> bool {
+        if let Some(round_id) = self.first_non_finalized_round_id.checked_sub(1) {
+            self.accepted_switch_block(round_id)
```
If I understand correctly, after we propose a block that should become a switch block, we start sending dummy proposals, right? And it is possible that the switch block only gets finalized once a later dummy proposal gets finalized (because it is possible that WBA for the switch block round won't terminate, for example). So in such a case we wouldn't detect that a switch block has been finalized, would we?

Should this say `self.accepted_switch_block(round_id) || self.accepted_dummy_proposal(round_id)`?

EDIT: It seems that `accepted_switch_block` would also return `true` for a dummy proposal, so this should be fine after all (unless we want to change the semantics of `accepted_switch_block` to be more in line with what the name suggests :wink:).

fizyk20 commented on node/src/components/consensus/protocols/zug.rs:
```diff
@@ -0,0 +1,2218 @@
+        }
+        let ancestor_values = match maybe_parent_round_id {
+            Some(parent_round_id)
+                if self.accepted_switch_block(parent_round_id)
+                    || self.accepted_dummy_proposal(parent_round_id) =>
```
This might be a redundant check, as `accepted_switch_block` would return true for a dummy proposal, if I understand correctly. (Not sure if it should be doing that, though.)

afck replied:
I agree, it shouldn't; I'll fix `accepted_switch_block` and double-check the call sites.

afck committed:
Address review comments.

fizyk20 commented on node/src/components/consensus/highway_core/finality_detector/rewards.rs:
```diff
@@ -136,7 +136,9 @@ fn round_participation<'a, C: Context>(
     };
     maybe_unit.map_or(RoundParticipation::No, |(vh, unit)| {
-        if unit.round_exp > r_id.trailing_zeros() {
+        // Round length is not 0: It is computed as 2^round_exp from a valid WireUnit.
```
It's `2^round_exp * min_round_length`, isn't it?

fizyk20 commented on node/src/components/consensus/protocols/zug.rs:
```diff
@@ -0,0 +1,2218 @@
+
+    /// Adds a signed message content to the state, but does not call `update` and does not detect
+    /// faults. Will not add the content of the message if the WAL is failing to write or if the
+    /// message is sent from a known faulty validator.
```
Is the comment correct? It seems that even if the WAL fails, we'll still call `add_content_no_wal`, no?

afck replied:
I inlined the function in its one call site and renamed `add_content_no_wal` to `add_content`.

fizyk20 commented on node/src/components/consensus/protocols/zug.rs:
```diff
@@ -0,0 +1,2218 @@
+    }
+
+    /// Returns the accepted value from the given round and all its ancestors, or `None` if there is
+    /// no accepted value in that round yet.
```
It also returns `None` if any of the ancestor rounds turns out not to have an accepted value, doesn't it?

afck replied:
Actually, this method only has one call site anymore anyway, so I'll inline it and rename `add_content_no_wal` to `add_content`.

afck committed:
Handle outcomes when reading Zug protocol state from disk.

afck committed:
Address a few TODOs.

fizyk20 reviewed (approved):
Reviewed Zug as well - a few minor nitpicks, no blockers :slightly_smiling_face:

afck committed:
Address more comments.

afck committed:
Increase run_participating_network timeout.

This further increases the timeout for run_participating_network, which
is flaky in CI but passes locally.

Also allows configuring more validators in the participating reactor
tests.

afck commented:
bors merge

afck commented:
Trying to debug the failing upgrade test. It says not all nodes were up after the upgrade, but they all have logs up to roughly the same timestamp. However, it looks like node 4 never `finished joining` after the upgrade.
~Also, the test uses Highway before and Zug only after the upgrade!~

This may still all be a timing issue, because I changed the minimum block time to 4 seconds, whereas the minimum Highway round length used to be 16. I'll change them back ~, and use Zug everywhere~.

(Ah, the test uses an older binary before the upgrade, so of course it uses Highway!)

afck committed:
Increase minimum block time.

This sets the minimum block time back to what it was based on the
minimum round length configured for Highway.

afck commented:
bors try

afck committed:
Move common types to consensus::utils.

afck committed:
Use different timeouts in run_participating_network.

That way, the two `settle_on` calls can be easily distinguished if it
fails.

afck committed:
Apply rustfmt.

afck committed:
In tests, check for >= era instead of ==.

afck commented:
Closing in favor of #3419.
